File: plugins/a2a-protocol/skills/a2a-executor-patterns/templates/async-executor.py
"""
Async A2A Executor - Long-Running Pattern

Use for: Tasks that take time and need status updates
Examples: LLM inference, file processing, data analysis
"""

from dataclasses import dataclass, field
from typing import Any, Dict, Optional, Callable
from datetime import datetime
from enum import Enum
import asyncio
import json


class TaskStatus(str, Enum):
    """Task status values"""
    PENDING = 'pending'
    RUNNING = 'running'
    COMPLETED = 'completed'
    FAILED = 'failed'


# Task and Result Classes
@dataclass
class A2ATask:
    """A2A Task definition"""
    id: str
    type: str
    parameters: Dict[str, Any]
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class A2AResult:
    """A2A Result definition"""
    task_id: str
    status: TaskStatus
    result: Optional[Any] = None
    error: Optional[str] = None
    progress: int = 0
    metadata: Optional[Dict[str, Any]] = None


@dataclass
class TaskState:
    """Internal task state"""
    task: A2ATask
    status: TaskStatus
    result: Optional[Any] = None
    error: Optional[str] = None
    progress: int = 0
    start_time: datetime = field(default_factory=datetime.utcnow)
    end_time: Optional[datetime] = None


# Task Store
class TaskStore:
    """In-memory task state storage"""

    def __init__(self):
        self.tasks: Dict[str, TaskState] = {}

    def create_task(self, task: A2ATask) -> TaskState:
        """Create new task state"""
        state = TaskState(
            task=task,
            status=TaskStatus.PENDING,
            progress=0
        )
        self.tasks[task.id] = state
        return state

    def update_task(self, task_id: str, **updates) -> None:
        """Update task state"""
        if task_id in self.tasks:
            state = self.tasks[task_id]
            for key, value in updates.items():
                setattr(state, key, value)

    def get_task(self, task_id: str) -> Optional[TaskState]:
        """Get task state"""
        return self.tasks.get(task_id)

    def delete_task(self, task_id: str) -> None:
        """Delete task state"""
        if task_id in self.tasks:
            del self.tasks[task_id]


# Global task store
task_store = TaskStore()


# Async Executor
async def submit_task(task: A2ATask) -> A2AResult:
    """Submit task for async execution"""
    # Create task state
    task_store.create_task(task)

    # Start async execution (don't await)
    asyncio.create_task(execute_task_async(task))

    # Return immediate response
    return A2AResult(
        task_id=task.id,
        status=TaskStatus.PENDING,
        progress=0,
        metadata={
            'submitted_at': datetime.utcnow().isoformat()
        }
    )


async def execute_task_async(task: A2ATask) -> None:
    """Execute task asynchronously"""
    try:
        # Update to running
        task_store.update_task(
            task.id,
            status=TaskStatus.RUNNING,
            progress=10
        )

        # Execute task
        async def on_progress(progress: int):
            task_store.update_task(task.id, progress=progress)

        result = await process_long_running_task(task, on_progress)

        # Update to completed
        task_store.update_task(
            task.id,
            status=TaskStatus.COMPLETED,
            result=result,
            progress=100,
            end_time=datetime.utcnow()
        )

        # Optional: Send callback notification
        await send_callback(task, result)

    except Exception as error:
        task_store.update_task(
            task.id,
            status=TaskStatus.FAILED,
            error=str(error),
            end_time=datetime.utcnow()
        )


async def process_long_running_task(
    task: A2ATask,
    on_progress: Callable[[int], None]
) -> Any:
    """Process long-running task with progress updates"""
    # Simulate long-running task
    steps = 10
    for i in range(steps):
        await asyncio.sleep(0.5)
        await on_progress(10 + (i + 1) * 8)  # Progress from 10% to 90%

    # Return result based on task type
    if task.type == 'llm-inference':
        return await run_llm_inference(task.parameters)

    elif task.type == 'file-processing':
        return await process_file(task.parameters)

    elif task.type == 'data-analysis':
        return await analyze_data(task.parameters)

    else:
        raise ValueError(f'Unsupported task type: {task.type}')


# Get Task Status
def get_task_status(task_id: str) -> Optional[A2AResult]:
    """Get current task status"""
    state = task_store.get_task(task_id)
    if not state:
        return None

    return A2AResult(
        task_id=task_id,
        status=state.status,
        result=state.result,
        error=state.error,
        progress=state.progress,
        metadata={
            'start_time': state.start_time.isoformat(),
            'end_time': state.end_time.isoformat() if state.end_time else None
        }
    )


# Cancel Task
async def cancel_task(task_id: str) -> bool:
    """Cancel running task"""
    state = task_store.get_task(task_id)
    if not state or state.status in [TaskStatus.COMPLETED, TaskStatus.FAILED]:
        return False

    task_store.update_task(
        task_id,
        status=TaskStatus.FAILED,
        error='Task cancelled',
        end_time=datetime.utcnow()
    )

    return True


# Example Task Processors
async def run_llm_inference(parameters: Dict[str, Any]) -> Any:
    """Implement LLM inference"""
    prompt = parameters.get('prompt')
    model = parameters.get('model')
    # Call LLM API here
    return {
        'response': 'Generated response',
        'model': model,
        'tokens': 100
    }


async def process_file(parameters: Dict[str, Any]) -> Any:
    """Implement file processing"""
    file_url = parameters.get('fileUrl')
    operation = parameters.get('operation')
    # Process file here
    return {
        'processed': True,
        'operation': operation,
        'file_url': file_url
    }


async def analyze_data(parameters: Dict[str, Any]) -> Any:
    """Implement data analysis"""
    data = parameters.get('data')
    analysis_type = parameters.get('analysisType')
    # Analyze data here
    return {
        'analysis': 'Results',
        'type': analysis_type
    }


# Optional: Callback notification
async def send_callback(task: A2ATask, result: Any) -> None:
    """Send callback notification"""
    if not task.metadata or 'callback_url' not in task.metadata:
        return

    callback_url = task.metadata['callback_url']

    try:
        # Use aiohttp or httpx for async HTTP
        # For this example, we'll just log
        print(f'Would send callback to: {callback_url}')
    except Exception as error:
        print(f'Callback failed: {error}')


# Example Usage
if __name__ == '__main__':
    async def main():
        example_task = A2ATask(
            id='async-task-001',
            type='llm-inference',
            parameters={
                'prompt': 'Explain quantum computing',
                'model': 'gpt-4'
            }
        )

        result = await submit_task(example_task)
        print('Task submitted:', json.dumps({
            'task_id': result.task_id,
            'status': result.status.value,
            'progress': result.progress
        }, indent=2))

        # Poll for status
        while True:
            await asyncio.sleep(1)
            status = get_task_status(example_task.id)
            if status:
                print('Status:', json.dumps({
                    'task_id': status.task_id,
                    'status': status.status.value,
                    'progress': status.progress
                }, indent=2))

                if status.status in [TaskStatus.COMPLETED, TaskStatus.FAILED]:
                    break

    asyncio.run(main())

File: plugins/a2a-protocol/skills/a2a-executor-patterns/templates/async-executor.ts
/**
 * Async A2A Executor - Long-Running Pattern
 *
 * Use for: Tasks that take time and need status updates
 * Examples: LLM inference, file processing, data analysis
 */

// Task and Result Types
interface A2ATask {
  id: string
  type: string
  parameters: Record<string, any>
  metadata?: Record<string, any>
}

interface A2AResult {
  taskId: string
  status: 'pending' | 'running' | 'completed' | 'failed'
  result?: any
  error?: string
  progress?: number
  metadata?: Record<string, any>
}

interface TaskState {
  task: A2ATask
  status: A2AResult['status']
  result?: any
  error?: string
  progress: number
  startTime: Date
  endTime?: Date
}

// Task Store
class TaskStore {
  private tasks = new Map<string, TaskState>()

  createTask(task: A2ATask): TaskState {
    const state: TaskState = {
      task,
      status: 'pending',
      progress: 0,
      startTime: new Date()
    }
    this.tasks.set(task.id, state)
    return state
  }

  updateTask(taskId: string, update: Partial<TaskState>): void {
    const state = this.tasks.get(taskId)
    if (state) {
      Object.assign(state, update)
    }
  }

  getTask(taskId: string): TaskState | undefined {
    return this.tasks.get(taskId)
  }

  deleteTask(taskId: string): void {
    this.tasks.delete(taskId)
  }
}

// Global task store
const taskStore = new TaskStore()

// Async Executor
async function submitTask(task: A2ATask): Promise<A2AResult> {
  // Create task state
  taskStore.createTask(task)

  // Start async execution (don't await)
  executeTaskAsync(task).catch(error => {
    console.error('Async execution error:', error)
    taskStore.updateTask(task.id, {
      status: 'failed',
      error: error.message,
      endTime: new Date()
    })
  })

  // Return immediate response
  return {
    taskId: task.id,
    status: 'pending',
    progress: 0,
    metadata: {
      submittedAt: new Date().toISOString()
    }
  }
}

async function executeTaskAsync(task: A2ATask): Promise<void> {
  try {
    // Update to running
    taskStore.updateTask(task.id, {
      status: 'running',
      progress: 10
    })

    // Execute task
    const result = await processLongRunningTask(task, (progress) => {
      taskStore.updateTask(task.id, { progress })
    })

    // Update to completed
    taskStore.updateTask(task.id, {
      status: 'completed',
      result,
      progress: 100,
      endTime: new Date()
    })

    // Optional: Send callback notification
    await sendCallback(task, result)
  } catch (error) {
    taskStore.updateTask(task.id, {
      status: 'failed',
      error: error instanceof Error ? error.message : 'Unknown error',
      endTime: new Date()
    })
  }
}

async function processLongRunningTask(
  task: A2ATask,
  onProgress: (progress: number) => void
): Promise<any> {
  // Simulate long-running task with progress updates
  const steps = 10
  for (let i = 0; i < steps; i++) {
    await new Promise(resolve => setTimeout(resolve, 500))
    onProgress(10 + (i + 1) * 8) // Progress from 10% to 90%
  }

  // Return result based on task type
  switch (task.type) {
    case 'llm-inference':
      return await runLLMInference(task.parameters)

    case 'file-processing':
      return await processFile(task.parameters)

    case 'data-analysis':
      return await analyzeData(task.parameters)

    default:
      throw new Error(`Unsupported task type: ${task.type}`)
  }
}

// Get Task Status
function getTaskStatus(taskId: string): A2AResult | null {
  const state = taskStore.getTask(taskId)
  if (!state) {
    return null
  }

  return {
    taskId,
    status: state.status,
    result: state.result,
    error: state.error,
    progress: state.progress,
    metadata: {
      startTime: state.startTime.toISOString(),
      endTime: state.endTime?.toISOString()
    }
  }
}

// Cancel Task
async function cancelTask(taskId: string): Promise<boolean> {
  const state = taskStore.getTask(taskId)
  if (!state || state.status === 'completed' || state.status === 'failed') {
    return false
  }

  taskStore.updateTask(taskId, {
    status: 'failed',
    error: 'Task cancelled',
    endTime: new Date()
  })

  return true
}

// Example Task Processors
async function runLLMInference(parameters: Record<string, any>): Promise<any> {
  // Implement LLM inference
  const { prompt, model } = parameters
  // Call LLM API here
  return {
    response: 'Generated response',
    model,
    tokens: 100
  }
}

async function processFile(parameters: Record<string, any>): Promise<any> {
  // Implement file processing
  const { fileUrl, operation } = parameters
  // Process file here
  return {
    processed: true,
    operation,
    fileUrl
  }
}

async function analyzeData(parameters: Record<string, any>): Promise<any> {
  // Implement data analysis
  const { data, analysisType } = parameters
  // Analyze data here
  return {
    analysis: 'Results',
    type: analysisType
  }
}

// Optional: Callback notification
async function sendCallback(task: A2ATask, result: any): Promise<void> {
  const callbackUrl = task.metadata?.callbackUrl
  if (!callbackUrl) return

  try {
    await fetch(callbackUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        taskId: task.id,
        status: 'completed',
        result
      })
    })
  } catch (error) {
    console.error('Callback failed:', error)
  }
}

// Export
export {
  A2ATask,
  A2AResult,
  submitTask,
  getTaskStatus,
  cancelTask,
  taskStore
}

// Example Usage
if (require.main === module) {
  const exampleTask: A2ATask = {
    id: 'async-task-001',
    type: 'llm-inference',
    parameters: {
      prompt: 'Explain quantum computing',
      model: 'gpt-4'
    }
  }

  submitTask(exampleTask)
    .then(result => {
      console.log('Task submitted:', result)

      // Poll for status
      const interval = setInterval(() => {
        const status = getTaskStatus(exampleTask.id)
        console.log('Status:', status)

        if (status?.status === 'completed' || status?.status === 'failed') {
          clearInterval(interval)
        }
      }, 1000)
    })
}
